compiler/modules/types/list.bal

```diff
@@ -118,25 +118,32 @@ function listSubtypeIsEmpty(Context cx, SubtypeData t) returns boolean {
     Bdd b = <Bdd>t;
     BddMemo? mm = cx.listMemo[b];
     BddMemo m;
+    Bdd[]? originalKeys = ();
     if mm == () {
         m = { bdd: b };
         cx.listMemo.add(m);
     }
     else {
         m = mm;
         boolean? res = m.isEmpty;
-        if res == () {
-            // we've got a loop
-            // XXX is this right???
-            return true;
-        }
-        else {
+        if res is boolean {
             return res;
         }
+        // we've got a loop
+        m.isEmpty = true;
+        originalKeys = cx.listMemo.keys();
     }
     boolean isEmpty = bddEvery(cx, b, (), (), listFormulaIsEmpty);
+    if originalKeys !is () && isEmpty != m.isEmpty {
```

Comment:
@jclark given that we need to clear some parts of the memo if we got our guess wrong in a loop I am not sure if this is any better than the [original solution](https://github.com/ballerina-platform/nballerina/issues/1065#issuecomment-1236024804).

Reply:
Why do we need this `originalKeys` stuff? I thought everything worked without that.

Reply:
Don't understand what you are doing with `originalKeys` at all.

Reply:
Your comment here doesn't have that https://github.com/ballerina-platform/nballerina/issues/1065#issuecomment-1236092188

Reply:
Can you please add a description of the proposed algorithm to #1065?

Reply:
Add https://github.com/ballerina-platform/nballerina/issues/1065#issuecomment-1236287740. This is the implementation (not sure this is the best way to do it) of the part in my original comment where I say `// Clean up if guess is wrong`